### PR TITLE
[5.1] Implement Schema.org Product type plugin

### PR DESCRIPTION
**Pull request by krileon**
### Summary of Changes
Implements Schema.org plugin for Product type.

### Testing Instructions
1. Ensure the Product schema plugin is installed (if switching to this branch be sure to use install discovery) and published
2. Edit any Article of your choosing in Content > Articles
3. Under the Schema tab select "Product" Schema type
4. Add whatever product information you would like and save the article
5. View the article on frontend to verify product schema has been output to the head

Your schema output on frontend should be for example similar to the below, which has already been ran through schema test tools and verified (non-critical issues will be reported for missing ratings and can add support for supplying ratings if desired, but seamed weird to hardcode that in).

image

### Actual result BEFORE applying this Pull Request
No Product Schema.org type available

### Expected result AFTER applying this Pull Request
Product Schema.org type available

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by sergeytolkachyov** [comment]
I have tested this item :white_check_mark: successfully on 0e7b2e45238a2bbccd01c6843e31f481afca8716<br /><br />Tested it successfull. Updated from 5.0.1 <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42710">issues.joomla.org/tracker/joomla-cms/42710</a>.</sub>

File: administrator/language/en-GB/plg_schemaorg_product.ini [hunk 1]
```diff
@@ -0,0 +1,39 @@
+; Joomla! Project
+; (C) 2024 Open Source Matters, Inc. <https://www.joomla.org>
+; License GNU General Public License version 2 or later; see LICENSE.txt
+; Note : All ini files need to be saved as UTF-8
+
+PLG_SCHEMAORG_PRODUCT="Schema.org - Product"
+PLG_SCHEMAORG_PRODUCT_DESCRIPTION_LABEL="More information on the Schema.org Product Type can be found at <a href=\"https://schema.org/Product\" target=\"_blank\" rel=\"noopener noreferrer\">https://schema.org/Product</a>"
+PLG_SCHEMAORG_PRODUCT_FIELD_ADDRESS_LABEL="Address"
```
Comment (brianteeman) [inline administrator/language/en-GB/plg_schemaorg_product.ini:8]: please alphasort the language strings

File: libraries/src/Extension/ExtensionHelper.php [hunk 2]
```diff
@@ -285,6 +285,7 @@ class ExtensionHelper
         ['plugin', 'organization', 'schemaorg', 0],
         ['plugin', 'person', 'schemaorg', 0],
         ['plugin', 'recipe', 'schemaorg', 0],
+        ['plugin', 'product', 'schemaorg', 0],
```
Comment (Quy) [inline libraries/src/Extension/ExtensionHelper.php]:
```suggestion
        ['plugin', 'person', 'schemaorg', 0],
        ['plugin', 'product', 'schemaorg', 0],
        ['plugin', 'recipe', 'schemaorg', 0],
```

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on d7048728674c1ed8a50d2b161542eb2d7103cf39<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42710">issues.joomla.org/tracker/joomla-cms/42710</a>.</sub>

File: plugins/schemaorg/product/forms/schemaorg.xml [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+									name="@type"
+									type="hidden"
+									default="Offer"
+								/>
+
+								<field
+									name="price"
+									type="text"
+									filter="string"
+									label="PLG_SCHEMAORG_PRODUCT_FIELD_VALUE_LABEL"
```
Comment (Quy) [inline plugins/schemaorg/product/forms/schemaorg.xml]: Please sort in the following order: name, type, label

Comment (krileon) [reply plugins/schemaorg/product/forms/schemaorg.xml]: I will fix it, but it was copy and pasted from jobposting to try and maintain exact code style. genericValue for example has label last in literally every schema type. So core code style isn't even following code style, which seams to be invented anew in every PR. Just an FYI. I feel like we're also getting into ridiculous levels of nitpicking here and it's a detriment to attracting contributors, just my 2 cents.

File: installation/sql/mysql/base.sql [hunk 4]
```diff
@@ -341,6 +341,7 @@ INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`,
 (0, 'plg_schemaorg_organization', 'plugin', 'organization', 'schemaorg', 0, 1, 1, 0, 0, '', '{}', '', 6, 0),
 (0, 'plg_schemaorg_person', 'plugin', 'person', 'schemaorg', 0, 1, 1, 0, 0, '', '{}', '', 7, 0),
 (0, 'plg_schemaorg_recipe', 'plugin', 'recipe', 'schemaorg', 0, 1, 1, 0, 0, '', '{}', '', 8, 0),
+(0, 'plg_schemaorg_product', 'plugin', 'product', 'schemaorg', 0, 1, 1, 0, 0, '', '{}', '', 9, 0),
```
Comment (richard67) [inline installation/sql/mysql/base.sql]:
```suggestion
(0, 'plg_schemaorg_product', 'plugin', 'product', 'schemaorg', 0, 1, 1, 0, 0, '', '{}', '', 8, 0),
(0, 'plg_schemaorg_recipe', 'plugin', 'recipe', 'schemaorg', 0, 1, 1, 0, 0, '', '{}', '', 9, 0),
```
Alpha ordering the plugins of the same type (folder).

File: installation/sql/postgresql/base.sql [hunk 5]
```diff
@@ -347,6 +347,7 @@ INSERT INTO "#__extensions" ("package_id", "name", "type", "element", "folder",
 (0, 'plg_schemaorg_organization', 'plugin', 'organization', 'schemaorg', 0, 1, 1, 0, 0, '', '{}', '', 6, 0),
 (0, 'plg_schemaorg_person', 'plugin', 'person', 'schemaorg', 0, 1, 1, 0, 0, '', '{}', '', 7, 0),
 (0, 'plg_schemaorg_recipe', 'plugin', 'recipe', 'schemaorg', 0, 1, 1, 0, 0, '', '{}', '', 8, 0),
+(0, 'plg_schemaorg_product', 'plugin', 'product', 'schemaorg', 0, 1, 1, 0, 0, '', '{}', '', 9, 0),
```
Comment (richard67) [inline installation/sql/postgresql/base.sql]:
```suggestion
(0, 'plg_schemaorg_product', 'plugin', 'product', 'schemaorg', 0, 1, 1, 0, 0, '', '{}', '', 8, 0),
(0, 'plg_schemaorg_recipe', 'plugin', 'recipe', 'schemaorg', 0, 1, 1, 0, 0, '', '{}', '', 9, 0),
```
Alpha ordering the plugins of the same type (folder).

**Comment by sergeytolkachyov** [comment]
I have tested this item :white_check_mark: successfully on 82d21228d2a8bcf40c1ae69c33e9eccd51a201ac<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42710">issues.joomla.org/tracker/joomla-cms/42710</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on 82d21228d2a8bcf40c1ae69c33e9eccd51a201ac<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42710">issues.joomla.org/tracker/joomla-cms/42710</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42710">issues.joomla.org/tracker/joomla-cms/42710</a>.</sub>

**Comment by krileon** [comment]
I don't understand what lead to your decision not to merge. At it's most basic it's just another XML form and there's nothing here to really maintain. There are a significant amount of eCommerce sites using Joomla that could massively benefit from the Product schema.

The Schema.org plugin isn't just for articles. It was designed to work with any extension. So yeah it may not make total sense adding it to an Article, but the Schema.org plugin is far more than just for that. I can't see how this is less useful than Event or JobPosting in that regard.

Are these discussions public so I can read what the concerns were?

**Comment by sergeytolkachyov** [comment]
> We discussed this PR in the weekliy maintainers meeting and made the decsion not to merge this PR. It's not question that this Schema.org definition can be useful in some cases but this will be true for many others too. We need to draw a line what goes into core. The tools are there, we have an example how to implement this and that should be enough.
> 
> Thanks for the proposal and for working on this.

Why?? On Joomla, small catalogs of goods and services are very often made using article overrides, where this plugin is simply necessary. Why can a Job or Event be included in the core, but not a Product?

**Comment by rdeutz** [comment]
> I don't understand what lead to your decision not to merge. At it's most basic it's just another XML form and there's nothing here to really maintain. There are a significant amount of eCommerce sites using Joomla that could massively benefit from the Product schema.

The aim for production is to have lighter core. What we want to do is to give people the tools to extent Joomla! We want to have an example that people can see how they can extent the provided set of functionality. We have the Schema.org plugins and anyone can make a plugin to support what makes sense on his website. 

I understand that closing a PR is frustrating for the author. But as I said we need to draw a line. I am going to ask the JED team if they can make a category for Schema.org plugins so that the work here can find a place to help others.

**Comment by brianteeman** [comment]
If this was applied universally it would be understandable but its not

**Comment by laoneo** [comment]
This plugin would be a great benefit for the JED. Please post it there and I'm sure you will get a lot of appreciation.

**Comment by SonkinS** [comment]
Catalogs on Joomla are made very often, perhaps more often than events! For some, product support would be decisive when choosing CMS...

**Comment by krileon** [comment]
Features are added constantly. Schema.org feature already exists. This is simply adding 1 more XML form to it. The Schema doesn't change randomly or constantly there's nothing further to maintain here. Additionally it's a Schema.org plugin. It can be turned off or even uninstalled if someone doesn't want it.

eCommerce is important. Especially now more than ever. Drawing the line at eCommerce? This concerns me for Joomla's future to be honest. These are basic SEO features that every business should have. eCommerce capabilities are often the leading reason for someone to pick a specific CMS. Something as simple as a Product schema makes a big difference for SEO.

I cannot stress how silly decisions like this are going to continue to drive away contributors. It's as if we're beholden to a shadow council. This is an open source project. Your discussions should be public. Whether here in the ticket, the discussions section, the Joomla forums, or something similar to PHP internals. The lack of transparency is concerning.

**Comment by alikon** [comment]
from my personal point of view, 
i  fully agree with @laoneo 
>This plugin would be a great benefit for the JED. Please post it there and I'm sure you will get a lot of appreciation.

in this way your hard work would be not lost, and the whole community would benefit of it

p.s
core or not core sometimes are like "quicksand"

**Comment by laoneo** [comment]
@krileon what prevents you to put the plugin into the Joomla extension directory? When you install Joomla, then it has already over 240 extensions installed. So it is normal that we maintainers want to keep that number low. You contribute it, but we maintainers need to take care of it as soon as it is merged. As we are only a handful of people, we have to be very careful what we put under our umbrella.

It is not that we do not appreciate your contribution, but we appreciate it as well when you put it into the JED.

**Comment by rdeutz** [comment]
A JED category for Schema Plugins is created. Thanks to the JED team for doing it so quickly!

https://extensions.joomla.org/category/core-enhancements/schema-plugins/

**Comment by krileon** [comment]
> @krileon what prevents you to put the plugin into the Joomla extension directory?

I'm already dealing with over 30 projects and a customer base for them. All for Joomla. I've so much work to do as is I'm drowning. So if I'm to release this on my own accord I will do so through our own download channel to drive traffic to our products. I wanted to use what limited time I had and give back to Joomla instead. Now if there were maintenance to be had, which there really isn't with this plugin, then it would be spread out amongst a much larger community and isn't really an issue to be pinged in a ticket to come help.

I was hoping this was the year I'd get to seriously push some modern features into Joomla. Features that every site should have. I'm currently working on a PR for OpenGraph Metadata support in core Joomla, but from the sounds of it I just shouldn't even bother. That's the kind of message you're sending to future contributors.

I've another PR for adding Schema.org support to menu items. I suppose I should also just preemptively close that was well. You're making it very unclear where the line is to be drawn and because you discussed all of this internally not a single person that contributes has any clue why you make the decisions you make.

> When you install Joomla, then it has already over 240 extensions installed. So it is normal that we maintainers want to keep that number low.

Then stop accepting new features all together if this is the argument or should I just not make this a plugin and instead hardcode in the schema type? Is the issue here features or that it's a separate extension? I just am not going to accept this as the reason to reject adding an XML file for an additional Schema type. Again, the Schema is not ever changing. There's nothing to maintain here. I would completely understand hesitance to just blindly accept a new full blown component, but this just isn't that.

Maybe another suggestion for the JED is GitHub support. Point a JED listing to a GitHub project and it can be packaged and released automatically. Then create a Joomla project for housing official-unofficial extensions that the maintenance team will have no oversite for and would be entirely community driven extensions.

> You contribute it, but we maintainers need to take care of it as soon as it is merged. As we are only a handful of people, we have to be very careful what we put under our umbrella.

I would personally say the maintainers are everyone here that contributes to this project. We will be using this schema type on our very own website, which has existed since Mambo, so it will be maintained (there's nothing to maintain as it's a standardized Schema) at the very least by those that use it (us).

If you want more contributors then stop being so unaccepting of change and pushing contributors away from this project. You're creating your own loop of less and less contributors. There's a serious disconnect going on between those who want to contribute and the maintainers team.

At any rate the decision has been made and I've said my peace. Best of luck.

**Comment by sandewt** [comment]
>At any rate the decision has been made and I've said my peace. Best of luck.

Sometimes things don't go the way you would like. And that can be disappointing, especially if you see the usefulness of it and have put in the necessary time. Your contribution is greatly appreciated, given the many positive responses. With your thorough knowledge of Joomla, you can still make many valuable contributions.

**Pull request closed without merging**